src/Catalyst.Core.Modules.Dfs/BlockExchange/BitSwapService.cs

```diff
@@ -0,0 +1,557 @@
+
+        /// <summary>
+        ///   Creates a new instance of the <see cref="BitSwapService"/> class.
+        /// </summary>
+        public BitSwapService()
+        {
+            Protocols = new IBitswapProtocol[]
+            {
+                new Bitswap11 {BitswapService = this}
+            };
+        }
+
+        /// <summary>
+        ///   Provides access to other peers.
+        /// </summary>
+        public SwarmService SwarmService { get; set; }
+
+        /// <summary>
+        ///   Provides access to blocks of data.
+        /// </summary>
+        public IBlockApi BlockService { get; set; }
+
+        /// <summary>
+        ///   Statistics on the bitswap component.
+        /// </summary>
+        /// <seealso cref="IStatsApi"/>
+        public BitswapData Statistics
+        {
+            get
+            {
+                return new BitswapData
+                {
+                    BlocksReceived = _blocksReceived,
+                    BlocksSent = _blocksSent,
+                    DataReceived = _dataReceived,
+                    DataSent = _dataSent,
+                    DupBlksReceived = _dupBlksReceived,
+                    DupDataReceived = _dupDataReceived,
+                    ProvideBufLen = 0, // TODO: Unknown meaning
+                    Peers = SwarmService.KnownPeers.Select(p => p.Id),
+                    Wantlist = _wants.Keys
+                };
+            }
+        }
+
+        /// <summary>
+        ///   Gets the bitswap ledger for the specified peer.
+        /// </summary>
+        /// <param name="peer">
+        ///   The peer to get information on.  If the peer is unknown, then a ledger
+        ///   with zeros is returned.
+        /// </param>
+        /// <returns>
+        ///   Statistics on the bitswap blocks exchanged with the peer.
+        /// </returns>
+        /// <seealso cref="IBitSwapApi.GetBitSwapLedger"/>
+        public BitswapLedger PeerLedger(Peer peer)
+        {
+            return _peerLedgers.TryGetValue(peer, out var ledger) ? ledger : new BitswapLedger {Peer = peer};
+        }
+
+        /// <summary>
+        ///   Raised when a blocked is needed.
+        /// </summary>
+        /// <remarks>
+        ///   Only raised when a block is first requested.
+        /// </remarks>
+        public event EventHandler<CidEventArgs> BlockNeeded;
+
+        /// <inheritdoc />
+        public Task StartAsync()
+        {
+            _log.Debug("Starting");
+
+            foreach (var protocol in Protocols)
+            {
+                SwarmService.AddProtocol(protocol);
+            }
+
+            SwarmService.ConnectionEstablished += Swarm_ConnectionEstablished;
+
+            // TODO: clear the stats.
+            _peerLedgers.Clear();
+
+            return Task.CompletedTask;
+        }
+
+        // When a connection is established
+        // (1) Send the local peer's want list to the remote
+#pragma warning disable VSTHRD100 // Avoid async void methods
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Do not suppress issues.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)